pkg/bucket/gcs.go

```diff
@@ -0,0 +1,69 @@
+package bucket
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type gcsBucket struct {
+	name   string
+	client *storage.Client
+}
+
+func NewGCSBucket(ctx context.Context, name string, opts ...option.ClientOption) (Bucket, error) {
+	client, err := storage.NewClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &gcsBucket{
+		name:   name,
+		client: client,
+	}
+
+	return b, nil
+}
+
+func (b *gcsBucket) Put(ctx context.Context, key string, data io.Reader, objectSize int64) error {
+	bucket := b.client.Bucket(b.name)
+
+	w := bucket.Object(key).NewWriter(ctx)
+	w.ChunkSize = int(decidePartSize(objectSize))
```

Comment:
Does the GCS have a limitation on the maximum number of parts per upload?
If the GCS has a different limitation from Amazon S3, we need to adjust accordingly.

The Amazon S3 has the following limitation.
So MOCO adjusts the chunk size according to backup file size. Please reffer to https://github.com/cybozu-go/moco/issues/318
https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
>
Maximum number of parts per upload | 10,000
-- | --

Reply:
As far as I've researched, I haven't found any specific limits. However, specifying the chunk size involves a trade-off with memory, and it seems that it is already well tuned by default. I think it's okay to leave the chunk size specification to the default in the client library and remove the description, but what do you think?

> The Go client library uses a buffer size that's equal to the chunk size. The buffer size must be a multiple of 256 KiB (256 x 1024 bytes). Larger buffer sizes typically make uploads faster, but note that there's a tradeoff between speed and memory usage. If you're running several resumable uploads concurrently, you should set Writer.ChunkSize to a value that's smaller than 16 MiB to avoid memory bloat.

https://cloud.google.com/storage/docs/resumable-uploads#go

Reply:
I added a comment about chunk size.
https://github.com/cybozu-go/moco/pull/501/commits/816f1e48bcb749668bbeb9220790b339f8454366